src/goto-instrument/contracts/contracts.cpp

```diff
@@ -142,30 +157,44 @@ void code_contractst::check_apply_loop_contracts(
     converter.goto_convert(assumption, havoc_code, mode);
   }
 
-  // Temporary variables that store the decreases clause's values before and
-  // after the loop
-  symbol_exprt old_decreases_symbol =
-    new_tmp_symbol(
-      decreases_clause.type(), loop_head->source_location, function_name, mode)
-      .symbol_expr();
-  symbol_exprt new_decreases_symbol =
-    new_tmp_symbol(
-      decreases_clause.type(), loop_head->source_location, function_name, mode)
-      .symbol_expr();
+  // Create fresh temporary variables that store the multidimensional
+  // decreases clause's value before and after the loop
+  for(size_t i = 0; i < size_decreases_clasue; i++)
+  {
+    old_temporary_variables.push_back(new_tmp_symbol(
+                                        decreases_clause_vector[i].type(),
+                                        loop_head->source_location,
+                                        function_name,
+                                        mode)
+                                        .symbol_expr());
+    new_temporary_variables.push_back(new_tmp_symbol(
+                                        decreases_clause_vector[i].type(),
+                                        loop_head->source_location,
+                                        function_name,
+                                        mode)
+                                        .symbol_expr());
+  }
```

Comment:
```suggestion
  for(const auto &clause : decreases_clause.operands())
  {
    old_temporary_variables.push_back(new_tmp_symbol(
                                        clause.type(),
                                        loop_head->source_location,
                                        function_name,
                                        mode)
                                        .symbol_expr());
    new_temporary_variables.push_back(new_tmp_symbol(
                                        clause.type(),
                                        loop_head->source_location,
                                        function_name,
                                        mode)
                                        .symbol_expr());
  }
```
Same for all the other loops.

Reply:
Thanks a lot!